### DISCUSSION TITLE
I am seeing network requests in devtools after transitioning to wagmi

### DISCUSSION THREAD

**Discussion by ackvf** [wevm/wagmi#1281] [Q&A]
While trying to solve an issue with not working Goerli due to a rate limit (Error 429 Too many requests) https://github.com/wagmi-dev/wagmi/discussions/1146, we have realised that
1) the api key used in those requests doesn't even come from our code
2) why are we dealing with this anyway, "it has worked before" without wagmi
3) when using `web3-react`, there are no network requests at all. **What is going on?**

<img width="1037" alt="image" src="https://user-images.githubusercontent.com/2313018/201210830-4ea467d8-040d-429b-ae4c-8b56cff010fa.png">

![image](https://user-images.githubusercontent.com/2313018/201213336-eb350cbe-d1e2-4f0b-947d-463bdf442ac2.png)

**Comment by ackvf** [comment]
Well, we just discovered something. When we are using the `Provider` object, the requests are visible in Network tab and the RPC endpoints are those set by wagmi. When we use the `Signer` object instead, the requests disappear and the app just keeps working.

```
   const { data: signer } = useSigner()
-  const provider = useProvider()
+  const provider = signer.provider

// ... keep using `provider` in the rest of the app as if nothing changed

const chainService = new ChainService(provider, signer, chainId)
```

/edit okay, so there are some methods from Provider that are not present on the Signer, but I guess we can use provider for these rare calls. e.g. `provider.waitForTransaction`

/edit2 So, I should have seen this coming... 🤦‍♂️ Using `signer.provider` instead of `signer` fixed all issues.

previously 
```
   const { data: signer } = useSigner()
-  const provider = useProvider()
+  const provider = signer
```
now
```
   const { data: signer } = useSigner()
-  const provider = useProvider()
+  const provider = signer.provider
```

**Reply by johnpaulcas** [reply]
Hi @ackvf by doing this, it does mean that we can't leverage all the hooks that used by wagmi? and on this case is that provider will be like window.ethereum? thanks!

**Reply by ackvf** [reply]
Yes I believe that the `signer.provider` is the one injected by Metamask (window.ethereum). But I don't know the full extent of which provider is used in other hooks. The other provider, the actual `provider = useProvider()` is an adhoc provider - useful, for example, to query data from other chains without bothering the user with switching networks in metamask back and forth.

That said, the `signer.provider from useSigner` should be used by default, not the `provider from useProvider`.
Provider from `useProvider` should be used for rare cases when the request is not really related to the user or when you need multichain capabilities, etc.

https://github.com/wagmi-dev/wagmi/discussions/1300

**Reply by tmm** [reply]
The signer's provider isn't used at all at the moment. Curious why you think the signer's provider should be used by default instead of the wagmi client's RPC provider?

As a developer, using a RPC provider you "control" tends to be a lot more reliable than using wallet RPCs (that you don't control). Often, when using `configureChains`, connectors (e.g. `WalletConnectConnector`) end up using the same RPC endpoints as your provider anyway.

**Reply by johnpaulcas** [reply]
@ackvf cool, so basically doing the code below will prevent 429 on just providing alchemy or infura ids

```shell
const { data: signer } = useSigner()
const provider = signer.provider
const contract = useContract({
    address: contractAddress,
    abi: contractABI,
    signerOrProvider: provider
  })
```

thou, this will be like a raw call

**Reply by ackvf** [reply]
@tmm We have the capabilities to suggest an RPC endpoint to the user and it's up to them whether they accept it or decide to use their own RPC endpoints in case they know it's far better and more reliable.

Though, I am open to change this view. We have just migrated from web3-react to wagmi and this was the first major difference we found that caught us off guard.

/edit

Why it raised concerns at all was that prior to around 0.7.6 version, you were, by default, using a public Alchemy RPC endpoint which has reached its quota. When trying to solve this, we found that the architecture between wagmi and web3-react is widely different and we were puzzled why, even though we have different RPCs set in Metamask, wagmi would still call that public enpoint.

**Reply by johnpaulcas** [reply]
guess its a way to utilise the injected wallet RPC, but doing this, you will lost the hooks wagmi created

**Reply by ackvf** [reply]
@johnpaulcas the 429 issue was solved in latest versions of wagmi, I believe after 0.7.5. Just upgrade to latest and you should not have that issue any more. But yes, using that code snippet would use the definitions as provided by Metamask - configured by the user, so most likely working, bypassing whatever is configured by wagmi library. IIUC

**Reply by johnpaulcas** [reply]
I see, thanks man! @ackvf

**Reply by jxom** [reply]
Appriciate all the feedback @ackvf! Really appreciate it!

We have thought about supporting Wallet RPCs in the past, and have even considered adding it as an option when you `configureChain`s.

However, IMHO, I generally think it is not the greatest experience to defer the RPC endpoint to the users' wallet. Why?

1. It kicks down the "rate-limiting can" to the user – what if the wallet's RPC ends up getting rate-limited? It is definitely not guaranteed that every wallet user will be using their own RPC endpoints & API keys. In fact, most users would be using public RPC endpoints – which are more vulnerable to rate-limiting.

2. The dapp will have no longer have control over the RPC
  a. dapps will be putting their trust into a wallet RPC – what happens if that RPC becomes vulnerable to an attack (eg. sandwich attack)?
  b. they will be unable to debug RPC errors that may have been initiated from the application
  
In the short term, we highly recommend configuring chains with [multiple providers via `configureChains`](https://wagmi.sh/docs/providers/configuring-chains) – this allows the dapp to fall back to another provider when something goes wrong (ie. provider has reached their quota).
  
In the long term, we definitely want to have a more robust solution around dealing with multiple RPC providers, and minimizing side-effects that emit unneeded/redundant RPC requests – we are working on this!

**Reply by ackvf** [reply]
@jxom awesome!, thanks for giving me this point of view. Coming from a different library with a different mindset it was difficult to grasp this, but I think it does make sense. We will revisit how we use signers and providers in our app.

**Reply by vildantursic** [reply]
Hi @jxom,
Is there a possibility to add this option anyway and allow devs to willingly opt in for using users wallet RPC via some different configuration, or this will be totally out of the story for Wagmi?
